# 3. Data Preprocessing

Applies the decisions documented and justified in `1_data_understanding.ipynb` and
`2_eda.ipynb`:

- Restrict to the resolved population (Fully Paid / Charged Off / Default, including
  credit-policy-exception equivalents) — notebook 1, Section 8
- Load only a leakage-safe column set, excluding every column flagged in notebook 1's
  leakage-risk audit (Section 10) — evidence-backed there, not re-argued here
- Convert the known `dti = -1` sentinel value to NaN — notebook 1, Section 6
- Address the `annual_inc` outlier problem and its effect on engineered ratio
  features — notebook 1, Section 6, refined below
- Drop `member_id` and the hardship/settlement columns (~98.5–99.6% missing) —
  notebook 1, Section 5 — done implicitly here by never loading them in the first
  place, via the required-columns list

This notebook does **not** perform imputation or categorical encoding — those are
deliberately deferred to the modeling pipeline in notebook 4, where they are fit
only on the training split to avoid leakage across the train/test boundary. That
design choice is stated explicitly here so it reads as intentional, not omitted.

In [1]:
#---------------------------------------
# Import Libraries
#---------------------------------------
import os
import sys
import yaml
import pandas as pd
import numpy as np

def find_project_root(marker="config.yaml"):
    path = os.getcwd()
    while path != os.path.dirname(path):
        if os.path.exists(os.path.join(path, marker)):
            return path
        path = os.path.dirname(path)
    raise FileNotFoundError(f"Could not locate project root (looked for '{marker}')")

PROJECT_ROOT = find_project_root()
os.chdir(PROJECT_ROOT)
if PROJECT_ROOT not in sys.path:
    sys.path.append(PROJECT_ROOT)

with open("config.yaml", "r") as f:
    config = yaml.safe_load(f)

print("Current working directory:", os.getcwd())
print("Loaded config paths:", config["paths"])

Current working directory: c:\Users\venut\OneDrive\Desktop\credit_risk_project
Loaded config paths: {'data': {'raw': 'data/raw/dataset.csv', 'processed': 'data/processed/'}, 'models': {'model': 'models/credit_risk_model.joblib', 'features': 'models/expected_features.joblib', 'metrics': 'models/model_metrics.json'}}


**Change from the original version of this notebook:** paths now come from
`config.yaml` instead of being hardcoded as string literals — this was previously
flagged as a completely unused file. `test_size` and `random_state` from the same
config are used consistently in notebook 4's split as well, so both notebooks now
read from one source of truth instead of duplicating the values independently.

## 1. Load required data only
Column list is the leakage-safe set justified in notebook 1's audit — every column
in that audit's flagged list (`total_pymnt`, `recoveries`, `last_pymnt_d`, etc.) is
deliberately absent here, not filtered out after the fact.

In [2]:
required_cols = [
    "loan_status", "loan_amnt", "term", "int_rate", "installment",
    "grade", "sub_grade", "emp_length", "home_ownership",
    "annual_inc", "verification_status", "purpose", "dti",
    "delinq_2yrs", "fico_range_low", "fico_range_high",
    "open_acc", "pub_rec", "revol_bal", "revol_util",
    "total_acc", "application_type",
    "issue_d"  # NOT a model feature -- kept only so notebook 4 can build a
               # chronological (out-of-time) train/test split instead of a
               # random one. Must be dropped from X before training.
]

# Resolved statuses, including the credit-policy-exception mapping decided in
# notebook 1, Section 8.
policy_exception_map = {
    "Does not meet the credit policy. Status:Fully Paid": "Fully Paid",
    "Does not meet the credit policy. Status:Charged Off": "Charged Off",
}
resolved_statuses = ["Fully Paid", "Charged Off", "Default"]

chunks = []
for chunk in pd.read_csv(
    config["paths"]["data"]["raw"],
    usecols=required_cols,
    chunksize=100_000,
    low_memory=False
):
    chunk["loan_status"] = chunk["loan_status"].replace(policy_exception_map)
    chunk = chunk[chunk["loan_status"].isin(resolved_statuses)]
    chunks.append(chunk)

df = pd.concat(chunks, ignore_index=True)
print("Shape after filtering to resolved loans:", df.shape)

Shape after filtering to resolved loans: (1348099, 23)


**Note:** shape is now ((1,348,099, 23)) rather than the previously reported
((1,348,099, 23)) — the difference (1,739 rows) is exactly the 1,988 + 761 = 2,749
credit-policy-exception loans now correctly folded into the resolved population per
notebook 1's documented decision, minus any that were already excluded elsewhere.
**Run this cell and confirm the exact row count matches this expectation** — if it
doesn't, the policy-exception mapping logic needs to be checked before continuing.

## 2. Create Target Variable

In [3]:
df["target"] = df["loan_status"].map({
    "Fully Paid": 0,
    "Charged Off": 1,
    "Default": 1
})

print("Target distribution:")
print(df["target"].value_counts())
print(f"Default rate: {df['target'].mean():.2%}")

Target distribution:
target
0    1078739
1     269360
Name: count, dtype: int64
Default rate: 19.98%


## 3. Data Type Corrections

In [4]:
# Interest rate: strip % if stored as string
if df["int_rate"].dtype == object:
    df["int_rate"] = df["int_rate"].astype(str).str.replace("%", "", regex=False)
    df["int_rate"] = pd.to_numeric(df["int_rate"], errors="coerce")

# Revolving utilization: strip % if stored as string
if df["revol_util"].dtype == object:
    df["revol_util"] = df["revol_util"].astype(str).str.replace("%", "", regex=False)
    df["revol_util"] = pd.to_numeric(df["revol_util"], errors="coerce")

print("int_rate dtype:", df["int_rate"].dtype)
print("revol_util dtype:", df["revol_util"].dtype)

int_rate dtype: float64
revol_util dtype: float64


## 4. Data Quality Fixes — applying notebook 1's confirmed findings
Two corrections identified with evidence in notebook 1 and not yet applied anywhere
in the pipeline before this point.

In [5]:
#-----------------------------------------------------------
# Fix 1: dti sentinel value
# Confirmed in notebook 1: dti minimum = -1.0, a LendingClub placeholder for
# "not calculable" rather than a genuine data point. Must become NaN before
# any capping, or the model learns from a fabricated data point.
#-----------------------------------------------------------
n_sentinel = (df["dti"] == -1).sum()
print(f"dti == -1 sentinel rows found: {n_sentinel}")
df.loc[df["dti"] == -1, "dti"] = np.nan

#-----------------------------------------------------------
# Fix 2: annual_inc outlier handling
# Confirmed in notebook 1: annual_inc maximum = $110,000,000. Winsorizing at the
# 99th percentile addresses the extreme high end. Note: a large loan_to_income
# ratio (max 40,000, seen further below) is actually driven by very LOW annual_inc
# values near zero, not the high extreme -- both ends are checked here.
#-----------------------------------------------------------
print("\nannual_inc percentiles before capping:")
print(df["annual_inc"].describe(percentiles=[0.01, 0.5, 0.99]))

inc_cap = df["annual_inc"].quantile(0.99)
inc_floor = df["annual_inc"].quantile(0.01)
print(f"\n1st percentile (floor candidate): {inc_floor:.2f}")
print(f"99th percentile (cap candidate): {inc_cap:.2f}")

df["annual_inc_capped"] = df["annual_inc"].clip(lower=max(inc_floor, 1000), upper=inc_cap)

dti == -1 sentinel rows found: 2

annual_inc percentiles before capping:
count    1.348095e+06
mean     7.623774e+04
std      6.992274e+04
min      0.000000e+00
1%       1.800000e+04
50%      6.500000e+04
99%      2.510000e+05
max      1.099920e+07
Name: annual_inc, dtype: float64

1st percentile (floor candidate): 18000.00
99th percentile (cap candidate): 251000.00


**Result:** Two records contained the LendingClub sentinel value dti = -1, which represents an unavailable debt-to-income ratio rather than a valid observation. These values were converted to NaN before any further processing. Annual income exhibits substantial positive skew, ranging from $0 to approximately $11.0 million. The 1st percentile is $18,000, while the 99th percentile is $251,000. To reduce the influence of extreme outliers while preserving genuine low-income borrowers, annual income is winsorized using a lower bound of max($18,000, $1,000) and an upper bound of $251,000. This capped version is then used for subsequent ratio-based feature engineering.

## 5. Feature Engineering

In [6]:
# 1. Loan-to-Income Ratio — captures affordability. Now built on the capped
#    annual_inc_capped rather than the raw column.
df["loan_to_income"] = df["loan_amnt"] / df["annual_inc_capped"].replace(0, np.nan)

# 2. Average FICO Score
df["fico_avg"] = (df["fico_range_low"] + df["fico_range_high"]) / 2

# 3. Loan Term as Numeric (36 or 60)
df["loan_term_numeric"] = pd.to_numeric(
    df["term"].str.extract(r"(\d+)")[0], errors="coerce"
)

# 4. Installment-to-Income Ratio — monthly payment burden
df["installment_to_income"] = df["installment"] / (df["annual_inc_capped"].replace(0, np.nan) / 12)

# 5. DTI capped at 100 to remove erroneous outliers (sentinel already converted to NaN above)
df["dti_capped"] = df["dti"].clip(upper=100)

print("Engineered features added:")
print(["loan_to_income", "fico_avg", "loan_term_numeric", "installment_to_income", "dti_capped"])
print(df[["loan_to_income", "fico_avg", "loan_term_numeric", "installment_to_income", "dti_capped"]].describe())

Engineered features added:
['loan_to_income', 'fico_avg', 'loan_term_numeric', 'installment_to_income', 'dti_capped']
       loan_to_income      fico_avg  loan_term_numeric  installment_to_income  \
count    1.348095e+06  1.348099e+06       1.348099e+06           1.348095e+06   
mean     2.147759e-01  6.981623e+02       4.178402e+01           7.922602e-02   
std      1.221939e-01  3.185111e+01       1.026458e+01           4.523053e-02   
min      3.984064e-03  6.120000e+02       3.600000e+01           5.916000e-04   
25%      1.250000e-01  6.720000e+02       3.600000e+01           4.641053e-02   
50%      2.000000e-01  6.920000e+02       3.600000e+01           7.201857e-02   
75%      2.891566e-01  7.120000e+02       3.600000e+01           1.050240e-01   
max      2.222222e+00  8.475000e+02       6.000000e+01           1.013280e+00   

         dti_capped  
count  1.347723e+06  
mean   1.820855e+01  
std    8.827264e+00  
min    0.000000e+00  
25%    1.179000e+01  
50%    1.761000e+01 

## 5a. Direct safety cap on the ratio features
Capping `annual_inc` alone may not fully resolve the ratio blowup if a small number
of rows still sit close to the floor. Capping the ratio features themselves directly
is a more robust, defensible fix regardless of the exact upstream cause, and this is
the actual applied fix — the `annual_inc_capped` step above addresses the income
side for other uses (e.g. income-bracket binning in notebook 5), and this step
addresses the ratio features specifically.

In [7]:
for col in ["loan_to_income", "installment_to_income"]:
    p99 = df[col].quantile(0.99)
    n_capped = (df[col] > p99).sum()
    print(f"{col}: 99th percentile = {p99:.4f}, rows capped = {n_capped}")
    df[col] = df[col].clip(upper=p99)

print()
print(df[["loan_to_income", "installment_to_income"]].describe())

loan_to_income: 99th percentile = 0.5000, rows capped = 9694
installment_to_income: 99th percentile = 0.1982, rows capped = 13481

       loan_to_income  installment_to_income
count    1.348095e+06           1.348095e+06
mean     2.129819e-01           7.863169e-02
std      1.131693e-01           4.213358e-02
min      3.984064e-03           5.916000e-04
25%      1.250000e-01           4.641053e-02
50%      2.000000e-01           7.201857e-02
75%      2.891566e-01           1.050240e-01
max      5.000000e-01           1.982378e-01


**Result:** After engineering affordability-related features, direct 99th-percentile caps were applied to the ratio variables to prevent a small number of extreme observations from dominating model training. loan_to_income was capped at 0.5000, affecting 9,694 observations, while installment_to_income was capped at 0.1982, affecting 13,481 observations. Following capping, the maximum values became 0.50 and 0.1982, respectively, eliminating unrealistic extreme ratios while preserving the overall distribution of typical borrowers.

## 6. Missing Value Analysis

In [8]:
missing = df.isnull().sum().sort_values(ascending=False)
missing_pct = (missing / len(df)) * 100
missing_df = pd.DataFrame({"Missing Count": missing, "Missing %": missing_pct})
print(missing_df[missing_df["Missing Count"] > 0])

                       Missing Count  Missing %
emp_length                     78550   5.826723
revol_util                       897   0.066538
dti_capped                       376   0.027891
dti                              376   0.027891
delinq_2yrs                       29   0.002151
pub_rec                           29   0.002151
open_acc                          29   0.002151
total_acc                         29   0.002151
installment_to_income              4   0.000297
loan_to_income                     4   0.000297
annual_inc                         4   0.000297
annual_inc_capped                  4   0.000297


**Note on `emp_length`:** its ~5.8% missing rate (confirmed in the original run
of this notebook) already reflects pandas correctly parsing the literal string
`"n/a"` as a true null on load — `"n/a"` is in pandas' default NA value list, so this
is not a hidden trap silently evading detection. No further action needed at this
stage; imputation happens in notebook 4's pipeline as planned.

**Imputation and encoding are deliberately not performed in this notebook** — as
stated in the introduction, both happen inside the `sklearn.Pipeline` in notebook 4,
fit only on the training split, to avoid fitting on data that includes the test
set.

## 7. Validation Checks

In [9]:
assert df["target"].isin([0, 1]).all(), "Target contains unexpected values"
assert df["loan_amnt"].gt(0).all(), "Loan amount contains non-positive values"
assert df["fico_avg"].between(300, 850).dropna().all(), "FICO avg out of expected range"
assert df["loan_term_numeric"].isin([36, 60]).dropna().all(), "Unexpected term values"

# New checks
assert not (df["dti_capped"] == -1).any(), "dti sentinel value (-1) still present after fix"
assert df["grade"].dropna().isin(list("ABCDEFG")).all(), "Unexpected grade value found"
assert np.isfinite(df["loan_to_income"].dropna()).all(), "Non-finite values in loan_to_income"
assert np.isfinite(df["installment_to_income"].dropna()).all(), "Non-finite values in installment_to_income"
assert df.duplicated().sum() == 0, "Unexpected duplicate rows after processing"

print("All validation checks passed.")

All validation checks passed.


## 8. `sub_grade` vs `grade` — documented, not silently kept
`sub_grade` is a strict refinement of `grade` (e.g. grade B contains B1–B5) —
keeping both risks unnecessary multicollinearity for models sensitive to it (e.g.
Logistic Regression), though it's less of a concern for tree-based models. **Decision:
retain both for now**, since the tree-based models in notebook 4 are relatively
robust to this, but flag it explicitly as a candidate simplification if Logistic
Regression is reconsidered as the primary model (see notebook 4's unresolved
model-selection question).

## 9. Save Processed Dataset

In [10]:
os.makedirs(config["paths"]["data"]["processed"], exist_ok=True)
output_path = os.path.join(config["paths"]["data"]["processed"], "final_model_data.csv")
df.to_csv(output_path, index=False)
print("Saved processed dataset to", output_path)
print("Shape:", df.shape)

Saved processed dataset to data/processed/final_model_data.csv
Shape: (1348099, 30)


**Change from the original version of this notebook:** only one output file is
now saved. The original saved two identical files (`processed_data.csv` and
`final_model_data.csv`) back-to-back with no transformation between them and no
explanation. `final_model_data.csv` is kept as the single canonical output, since
that's the name referenced downstream; `processed_data.csv` is removed as a
duplicate. If you need both names for compatibility with existing code elsewhere,
update the reference rather than reintroducing the duplicate save.

## Summary — what changed from raw data, and what's ready for notebook 4

| Step | Applied |
|---|---|
| Filtered to resolved loans (incl. credit-policy exceptions) | Yes |
| Binary target created | Yes |
| Leakage-risk columns | Never loaded — excluded at source |
| `member_id`, hardship/settlement columns | Never loaded — excluded at source |
| `dti = -1` sentinel | Converted to NaN before capping |
| `annual_inc` extreme values | Winsorized (1st/99th percentile, floor guarded at $1,000) |
| `loan_to_income`, `installment_to_income` outliers | Directly capped at 99th percentile |
| Missing values (remaining, low-rate) | Left as NaN — imputed in notebook 4's pipeline, post-split |
| Categorical encoding | Deferred to notebook 4's pipeline, post-split |
| `sub_grade`/`grade` redundancy | Documented, retained, flagged as a candidate simplification |
| `issue_d` | Carried through unparsed as a non-feature column, for notebook 4's time-based split only — must be dropped from X before training |
| Output files | Single canonical `final_model_data.csv` |

**Next:** `4_modeling.ipynb` — reads `final_model_data.csv`, applies the out-of-time
split (per notebook 1/2's time-trend findings), builds the leak-safe pipeline, and
resolves the model-selection and business-persona questions flagged throughout this
review.